From 29cc6bc899a2f551bf7705c1ebdf9211f81b8474 Mon Sep 17 00:00:00 2001
From: Balint Reczey <balint.reczey@canonical.com>
Date: Tue, 17 Dec 2019 16:48:13 +0100
Subject: [PATCH 5/6] Return only PAM_IGNORE or error from pam_motd

Follow-up for c81280b16e1831ab0bdd0383486c7e2d1eaf1b5e.
* modules/pam_motd/pam_motd.c: Return PAM_IGNORE if pam_putenv succeeds.
* modules/pam_motd/pam_motd.8.xml: Document additional possible return values of the module.
---
 modules/pam_motd/pam_motd.8.xml | 18 +++++++++++++++++-
 modules/pam_motd/pam_motd.c     |  2 +-
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/modules/pam_motd/pam_motd.8.xml b/modules/pam_motd/pam_motd.8.xml
index 798b9aec..8e88e744 100644
--- a/modules/pam_motd/pam_motd.8.xml
+++ b/modules/pam_motd/pam_motd.8.xml
@@ -98,11 +98,27 @@
   <refsect1 id='pam_motd-return_values'>
     <title>RETURN VALUES</title>
     <variablelist>
+      <varlistentry>
+        <term>PAM_ABORT</term>
+        <listitem>
+           <para>
+             Not all relevant data or options could be obtained.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term>PAM_BUF_ERR</term>
+        <listitem>
+           <para>
+              Memory buffer error.
+          </para>
+        </listitem>
+      </varlistentry>
       <varlistentry>
         <term>PAM_IGNORE</term>
         <listitem>
           <para>
-            This is the only return value of this module.
+            This is the default return value of this module.
           </para>
         </listitem>
       </varlistentry>
diff --git a/modules/pam_motd/pam_motd.c b/modules/pam_motd/pam_motd.c
index 546958f3..bbb25e62 100644
--- a/modules/pam_motd/pam_motd.c
+++ b/modules/pam_motd/pam_motd.c
@@ -234,7 +234,7 @@ int pam_sm_open_session(pam_handle_t *pamh, int flags,
 
     retval = pam_putenv(pamh, "MOTD_SHOWN=pam");
 
-    return retval;
+    return retval == PAM_SUCCESS ? PAM_IGNORE : retval;
 }
 
 /* end of module definition */
-- 
2.17.1

